### PR TITLE
Correct derivatives for methods called on NotADuckDualNumbers

### PR DESCRIPTION
**Pull request by lindsayad**
Previously the derivatives resulting from calling a method on a `NotADuckDualNumber` could be horribly wrong. Let's imagine a dual number with u = 1, du/dx = -1. Previously, f = u.norm() would result in a dual number with f = 1, df/dx = 1, when in fact the correct result is f = 1, df/dx = -1. 

These changes correctly apply the chain rule (df/dx = df/du * du/dx) to compute the derivatives; the results are verified by comparing to the same operations conducted on an "inner template", e.g. I check that the value and derivatives for `NotADuckDualNumber<VectorValue<double>>::norm` and `VectorValue<DualNumber>::norm` are the same.

In order to get the derivatives correct, a couple of methods in the `NotADuckDualNumber` specialization have to be added, example below:
```c++
  TensorValue<DualNumber<double, DuckNumberDerivatives>> inner_template() const
  {
    DualNumber<double, DuckNumberDerivatives> xx{this->value()(0, 0), _e0};
    DualNumber<double, DuckNumberDerivatives> xy{this->value()(0, 1), _e1};
    DualNumber<double, DuckNumberDerivatives> xz{this->value()(0, 2), _e2};
    DualNumber<double, DuckNumberDerivatives> yx{this->value()(1, 0), _e3};
    DualNumber<double, DuckNumberDerivatives> yy{this->value()(1, 1), _e4};
    DualNumber<double, DuckNumberDerivatives> yz{this->value()(2, 1), _e5};
    DualNumber<double, DuckNumberDerivatives> zx{this->value()(2, 0), _e6};
    DualNumber<double, DuckNumberDerivatives> zy{this->value()(2, 1), _e7};
    DualNumber<double, DuckNumberDerivatives> zz{this->value()(2, 2), _e8};
    return {xx, xy, xz, yx, yy, yz, zx, zy, zz};
  }

  NotADuckDualNumber<TensorValue<double>, ResizedDerivatives>
  convert_to_outer(const TensorValue<DualNumber<double, DuckNumberDerivatives>> & inner) const
  {
    NotADuckDualNumber<TensorValue<double>, ResizedDerivatives> nd = TensorValue<double>(inner(0, 0).value(),
                                                                               inner(0, 1).value(),
                                                                               inner(0, 2).value(),
                                                                               inner(1, 0).value(),
                                                                               inner(1, 1).value(),
                                                                               inner(1, 2).value(),
                                                                               inner(2, 0).value(),
                                                                               inner(2, 1).value(),
                                                                               inner(2, 2).value());
    for (unsigned di = 0; di < 9; ++di)
      for (unsigned i = 0; i < 3; ++i)
        for (unsigned j = 0; j < 3; ++j)
          nd.derivatives()[di](i, j) = inner(i, j).derivatives()[di];
    return nd;
  }

  const DualNumber<double, DuckNumberDerivatives> &
  convert_to_outer(const DualNumber<double, DuckNumberDerivatives> & inner) const
    {
      return inner;
    }
```
The successive application of `inner_template` and then `convert_to_outer` results in correct computation of df/du in the outer template form.

Now one might ask why go through all this effort to keep the outer template form, when in the end I'm relying on the inner template form to do the derivative calculations correctly...the answer is that it allows us to do things like this really easily:
```c++
template <typename T>
class MaterialProperty : public PropertyValue
{
public:
  :
  :
  :
  T & operator[](const unsigned int i) { return _value[i].value(); }
  :
  :
  :
};

template <typename T>
class ADMaterialPropertyObject : public MaterialProperty<T>
{
public:
  :
  :
  :
  typename MetaPhysicL::NDDualNumber<T, NumberArray<AD_MAX_DOFS_PER_ELEM, T>> &
  operator[](const unsigned int i)
  {
    return this->_value[i];
  }
  :
  :
  :
};
```
where the base `MaterialProperty`class is used when computing residuals and the derived `ADMaterialProperty` class is used when computing Jacobians. The trade-off for this ease of design and use is the additional computational expense of these `inner_template` and `conver_to_outer` calculations.

A lot of words here, but hopefully they're helpful for review.

### CODE REVIEW

File: src/numerics/include/metaphysicl/numberarray.h [hunk 1]
```diff
@@ -74,6 +74,11 @@ class NumberArray
     typedef NumberArray<N, T2> other;
   };
 
+  template <size_t N2>
+  struct resize {
+    typedef NumberArray<N2, T> other;
+  };
+
```
Comment (roystgnr) [inline src/numerics/include/metaphysicl/numberarray.h]: This looks useful; add it to numbervector.h too while you're at it?

**Comment by roystgnr** [comment]
Looks good at first pass, but this is going to take closer examination; yell at me if I don't get back to you by Monday.

**Comment by lindsayad** [comment]
Polite yell

**Comment by roystgnr** [comment]
Thanks; sorry for the delay!

**Comment by lindsayad** [comment]
@roystgnr could we create a `pr22_bootstrapped` tag? I could also create it; don't want to be making burdens for you

**Comment by roystgnr** [comment]
Is this minor enough for another pr tag?  I'd say fixing "could be horribly wrong" justifies calling the result 0.3.0.2 or 0.3.1.

If you're in a hurry you can create it, if not I'll find time tonight or tomorrow.

**Comment by lindsayad** [comment]
Not in a hurry...yes I agree that a new version release might be warranted

**Comment by roystgnr** [comment]
Check out 0.3.1 and make sure I didn't screw anything up?

**Pull request merged**